### PR TITLE
On-demand Banner and Icon Download

### LINKED ISSUE #3458: Banners are not automatically downloaded anymore
Since a few months, banners stopped from being automatically added to a new installed game, even tho the "View on Lutris.net" button points correctly to the game page.

### LINKED ISSUE #3957: Newly added Lutris games do not download their banners
### Bug description

Lutris's banner and icon download is full of holes; it tries to download banners and icons when adding games or searching, but its spotty and games are easily introduced with no banner. They just come out at grey boxes.  

This could be considered a dupe of #3458, but I've got a repro for you. It's not just "I'd like a better architecture"- the current approach fails in some cases. Not just obscure cases either. Read on!

### How to Reproduce

First, go to Lutris.net and sign it, pick a game with a banner that you don't have, and add it to your library.

Now, open Lutris, and click the reload button on the Lutris sidebar row. Wait for it to complete.

Now, view the Lutris games and search for your new game. Don't forget to show non-installed games.

Your new game should appear, but it has no banner.

### Expected behavior

The banner should appear on the newly added game.

I had hoped for Lutris to download the banners for games that need them if missing, no matter how you add the game.

### Log output

```shell
INFO     2021-12-29 10:10:48,924 [startup.check_driver:53]:Using NVIDIA drivers 470.86 for x86_64
INFO     2021-12-29 10:10:48,924 [startup.check_driver:57]:GPU: NVIDIA GeForce RTX 3060
INFO     2021-12-29 10:10:48,924 [startup.check_driver:73]:GPU: 10DE:2503 196E:1377 (nvidia drivers)
DEBUG    2021-12-29 10:10:49,110 [lutriswindow.update_store:451]:Showing 184 games
DEBUG    2021-12-29 10:10:51,240 [base.wipe_game_cache:265]:Deleting lutris cache /home/danj/.cache/lutris/lutris
DEBUG    2021-12-29 10:10:51,240 [base.wipe_game_cache:113]:Deleting games from service-games for lutris
DEBUG    2021-12-29 10:10:51,259 [http.get:81]:GET https://lutris.net/api/games/library/danieljohnson2
DEBUG    2021-12-29 10:10:54,195 [lutris.load:108]:Matching with already installed games
DEBUG    2021-12-29 10:10:54,466 [lutris.load:111]:Lutris games loaded
DEBUG    2021-12-29 10:10:54,472 [lutriswindow.update_store:451]:Showing 185 games
INFO     2021-12-29 10:10:58,466 [application.do_shutdown:631]:Shutting down Lutris
```

### System Information

```shell
[System]
OS:              Pop!_OS 21.10 impish
Arch:            x86_64
Kernel:          5.15.8-76051508-generic
Desktop:         GNOME
Display Server:  x11

[CPU]
Vendor:          AuthenticAMD
Model:           AMD Ryzen 7 5800X 8-Core Processor
Physical cores:  8
Logical cores:   16

[Memory]
RAM:             31.4 GB
Swap:            0.0 GB

[Graphics]
Vendor:          NVIDIA Corporation
OpenGL Renderer: NVIDIA GeForce RTX 3060/PCIe/SSE2
OpenGL Version:  4.6.0 NVIDIA 470.86
OpenGL Core:     4.6.0 NVIDIA 470.86
OpenGL ES:       OpenGL ES 3.2 NVIDIA 470.86
Vulkan:          Supported
```

### Media (optional)

Screenshot from 2021-12-29 10-15-45

### Checklist:

- [X] I'm not asking for support with a game or the wine runner.
- [X] I have followed the above mentioned guides and have all the graphics and wine dependencies installed.
- [X] I have checked for existing issues that describe my problem prior to opening this one.
- [X] I understand that improperly formatted bug reports may be closed without explanation.

### PR DESCRIPTION
**Pull request by danieljohnson2**
This is a big PR; @RobinFrcd had a much smaller PR for this (#3460); @strycore closed this with "Too stale and drifts too far away of future designs.". I have no idea what future designs you have in mind, but this at least is not so stale.

It's also more ambitious. This uses a cell-renderer to trigger downloading banners and icons- and no longer pre-downloads any banners or icons. This may mean *less* traffic on your server- when you do a search, only the banners you can see will download, not the others you haven't scrolled to you. Also, if the user uses icons instead of banners, icons download and banners do not.

Download failures will result in empty banner or icon files, according to the existing mechanism. This will prevent a re-download attempt for a few weeks.

In order to batch downloads a bit more, it delays start of the download for half a second. This means you can see the blank banner briefly- but I the UI should not lock up waiting for the server response, as it does not wait for the banner to be available.

Resolves #3957
Resolves #3458

### CODE REVIEW

File: lutris/services/egs.py [hunk 1]
```diff
@@ -142,9 +142,6 @@ class EpicGamesStoreService(OnlineService):
         "game_box": DieselGameBox,
         "box_tall": DieselGameBoxTall,
     }
-    extra_medias = {
-        "logo": DieselGameBoxLogo,
```
Comment (strycore) [inline lutris/services/egs.py:146]: when does this get downloaded now?

**Comment by strycore** [comment]
I'm tempted to merge this and see how it behaves on my lutris library of usual size.

**Comment by danieljohnson2** [comment]
Give it a shot, it's definitely better at actually getting things downloaded. I've been using it to fix my library when a banner is missing. Though these days there seem to be a lot of games with no banner; a website issue I think.

This PR *only* downloads media that the user views, when they view it. If there's a need to download anything else, that will need to be handled. But does Lutris ever use media without showing them? If so, how? Maybe that could trigger a download also. Or are we just downloading extra stuff to just have it?

**Comment by strycore** [comment]
What interests me is that my lutris library is really big and the UI has been getting really slow when I added the cover art format. If that PR also works for only showing images currently viewed, then it could be a big improvement on the performance.

**Comment by danieljohnson2** [comment]
This branch did seem to make reloading my (not so large) library faster, just by not downloading all the banners.

**Comment by strycore** [comment]
I'm not so much interested in speeding up the reloading. The client is sluggish for a normal use, this is what I want to fix.

**Comment by strycore** [comment]
I tried the branch locally. It does not speed up the UI, quite the opposite. The UI becomes unresponsive when media gets downloaded.
Switching between different sources is still very slow so that patch doesn't solve this issue, which was the thing I was hoping for.
I still think loading media on demand is the preferred way to go, but we can't ship this in this state. I do think a custom CellRenderer is the correct solution, it's just not solving the right problem at the moment.

**Comment by danieljohnson2** [comment]
This branch doesn't work anymore; too much drift.

**Pull request closed without merging**